### PR TITLE
Initial commit 💈

### PR DESCRIPTION
**Pull request by muuvmuuv**
First attempt creating the csscomb beautifier. Note that I couldn't get csscomb fetching the .csscomb.json file, so I had to include cosmiconfig to do so.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -29,14 +32,14 @@ bower_components
 # node-waf configuration
 .lock-wscript
 
-# Compiled binary addons (https://nodejs.org/api/addons.html)
+# Compiled binary addons (http://nodejs.org/api/addons.html)
```
Comment (Glavin001) [inline .gitignore]: No, we want `https`: https://nodejs.org/api/addons.html

File: .gitignore [hunk 2]
```diff
@@ -29,14 +32,14 @@ bower_components
...
 build/Release
 
 # Dependency directories
 node_modules/
 jspm_packages/
 
-# TypeScript v1 declaration files
+# Typescript v1 declaration files
```
Comment (Glavin001) [inline .gitignore]: No, it `TypeScript` not `Typescript`.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,32 @@
+env:
+  global:
+    - CC_TEST_REPORTER_ID=
```
Comment (Glavin001) [inline .travis.yml]: @szeck87 Are you able to get this ID?

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      package: "csscomb",
+    },
+  ],
+  options: {
+    CSS: options.CSS,
+    SCSS: options.CSS,
+    Sass: options.CSS,
+  },
+  resolveConfig: ({ filePath, projectPath, dependencies }) => {
+    if (!filePath) {
```
Comment (Glavin001) [inline src/index.ts]: Could use the `projectPath` if `filePath` is not set.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    const Comb = dependencies.get<NodeDependency>("CSScomb").package;
+    const explorer = cosmiconfig("csscomb", {
+      searchPlaces: [
+        ".csscomb.json",
+        "csscomb.json",
+        ".csscomb.js",
+        "csscomb.js",
+      ],
+    });
+    const cosmicFund = explorer.searchSync(filePath);
```
Comment (Glavin001) [inline src/index.ts]: Do not need to use the `sync` method, if there is an async method you can return a promise 👍 .

File: .travis.yml (hunk 3 above)
Comment (szeck87) [reply .travis.yml]: I need more access to this repo.

**Comment by muuvmuuv** [comment]
Oh I used the starter @szeck87 suggested. Seems many of those comments you made are from there. We should update the starter too. Otherwise these will appear again and again when others starting there. :) @Glavin001 

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by szeck87** [comment]
I updated the template's .gitignore file to use https.  Thanks!

File: .travis.yml (hunk 3 above)
Comment (Glavin001) [reply .travis.yml]: @szeck87 try now :).

Comment (szeck87) [reply .travis.yml]: b32edb620b73c1bd0e42e7a0157494c5841273f45cf1fae873375e018e6fa89d

Comment (Glavin001) [reply .travis.yml]: @muuvmuuv use the above ^.

File: package.json [hunk 6]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@unibeautify/beautifier-csscomb",
+  "version": "0.0.1",
+  "description": "csscomb beautifier for Unibeautify",
```
Comment (Glavin001) [inline package.json]: Change to `CSScomb`

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  ],
+  options: {
+    CSS: options.CSS,
+    SCSS: options.CSS,
+    Sass: options.CSS,
+  },
+  resolveConfig: ({ filePath, projectPath, dependencies }) => {
+    const pathToSearch = filePath ? filePath : projectPath;
+    if (!pathToSearch) {
+      return Promise.resolve({});
```
Comment (Glavin001) [inline src/index.ts]: Could add a test here later.

File: src/index.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        "csscomb.json",
+        ".csscomb.js",
+        "csscomb.js",
+      ],
+    });
+    return explorer
+      .search(pathToSearch)
+      .then(result => {
+        if (!result) {
+          return Promise.resolve({
```
Comment (Glavin001) [inline src/index.ts]: Could add a test here later.

File: src/index.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+  }: BeautifierBeautifyData) {
+    return new Promise<string>((resolve, reject) => {
+      const CSScomb = dependencies.get<NodeDependency>("CSScomb").package;
+      const config = (beautifierConfig && beautifierConfig.config) || {};
+      const comb = new CSScomb(config);
+      const result = comb.processString(text);
+      if (result) {
+        return resolve(result);
+      } else {
+        return resolve(text);
```
Comment (Glavin001) [inline src/index.ts]: @szeck87 We may want beautifiers to actually throw in these cases, so we know if something went wrong. I recall a lot of issues in Atom-Beautify where users are wondering whether a beautifier is working and we wouldn't want the beautifier to report a success when it was really a failure. Something to consider for later.

Comment (szeck87) [reply src/index.ts]: What would we throw though?  It's not an error since the beautifier works, there just isn't anything for it to do.

Comment (muuvmuuv) [reply src/index.ts]: We should return an object with the returning text and some additional info; did the text successfully beautified: true/false and an error message if there was one. So we could process them in plugins like VS Code to display some information as a notification or in the logs.

Comment (Glavin001) [reply src/index.ts]: @muuvmuuv  Either

```javascript
throw new Error("the error message");
```

or

```
return Promise.reject(new Error("the error message"));
```

should be sufficient.

**Comment by muuvmuuv** [comment]
@Glavin001 do you mean something like that? #db1feba

File: src/index.ts (hunk 9 above)
Comment (muuvmuuv) [reply src/index.ts]: @szeck87 I think:

- there will always be a result output even if there is nothing to do for the beautifier when a file is already fine
- no result would mean there has been no text processed that could be beautified for the engine -> so it has failed somewhere

I'll try to set up some tests for that

**Comment by Glavin001** [comment]
Looking good, @muuvmuuv ! Tests appear to be failing currently. I think this is a step in the right direction 😃 . Thanks!

**Comment by muuvmuuv** [comment]
@Glavin001 I can not work on this for the next week, maybe longer. Hopefully, I'll have time then to try some more tests (not quite easy, as it is no longer maintained)

**Comment by Glavin001** [comment]
Great work on this PR, @muuvmuuv ! I'll try to finish it off and publish soon 🎉 .

**Comment by Glavin001** [comment]
@muuvmuuv I fixed the failing tests in https://github.com/Unibeautify/beautifier-csscomb/commit/3f7422b72517320169ab7db542d20831b0c0bfd7 

Is there anything else that needed to be done? 😃

**Pull request merged**